corebuild/integration/ILLink.Tasks/ILLink.CrossGen.targets

```diff
@@ -0,0 +1,231 @@
+    </PropertyGroup>
+
+    <Message Text="$(CrossGenCommand)" />
+    <Exec Command="$(CrossGenCommand)" />
+
+  </Target>
+
+  <Target Name="_ComputeCrossGenInputs"
+          DependsOnTargets="_ComputeCrossGenPlatformAssemblies;_ComputeFilesToCrossGen" />
+
+  <!-- Compute the platform assembly paths, a parameter to crossgen
+       that lets it find dependencies of the input file. This needs to
+       include the path to the input file and its dependency closure,
+       including System.Private.CoreLib. -->
+  <Target Name="_ComputeCrossGenPlatformAssemblies"
+          DependsOnTargets="_ComputeManagedResolvedAssembliesForCrossGen">
+
+    <ItemGroup>
+      <_CrossGenPlatformAssemblies Include="@(_ManagedResolvedAssembliesForCrossGen)" />
+      <_CrossGenPlatformAssemblies Include="@(IntermediateAssembly)" />
+    </ItemGroup>
+
+  </Target>
+
+  <!-- _ManagedResolvedAssembliesForCrossGen includes the app's
+       managed dependency closure, including System.Private.CoreLib,
+       but not the app itself or resource assemblies. -->
+  <Target Name="_ComputeManagedResolvedAssembliesForCrossGen">
+    <ComputeManagedAssemblies Assemblies="@(ResolvedAssembliesToPublish)">
+      <Output TaskParameter="ManagedAssemblies" ItemName="_ManagedResolvedAssembliesForCrossGen" />
+    </ComputeManagedAssemblies>
+
+    <ItemGroup>
+      <_ManagedResolvedAssembliesForCrossGen Remove="@(_ManagedResolvedAssembliesForCrossGen->WithMetadataValue('AssetType', 'resources'))" />
+    </ItemGroup>
+  </Target>
+
+  <UsingTask TaskName="FilterByMetadata" AssemblyFile="$(LinkTaskDllPath)" />
+  <!-- This computes the default set of files that we want to be
+       crossgen'd. Some of these may already be R2R images, and these
+       will not be crossgen'd again. The default is to crossgen the
+       app and platform libraries. Defaults will be used only if
+       FilesToCrossGen hasn't been set elsewhere, allowing users and
+       other props/targets to select what will be crossgen'd. -->
+  <Target Name="_ComputeDefaultFilesToCrossGen"
+          DependsOnTargets="_ComputeManagedResolvedAssembliesForCrossGen;_ComputePlatformLibraries"
+          Condition=" '@(FilesToCrossGen->Count())' == '0' ">
+
+    <FilterByMetadata Items="@(_ManagedResolvedAssembliesForCrossGen)"
+                      MetadataName="Filename"
+                      MetadataValues="@(PlatformLibraries->'%(Filename)')">
+      <Output TaskParameter="FilteredItems" ItemName="_PlatformLibrariesForCrossGen" />
+    </FilterByMetadata>
+
+    <ItemGroup>
+      <FilesToCrossGen Include="@(IntermediateAssembly)" />
+      <FilesToCrossGen Include="@(_PlatformLibrariesForCrossGen)" />
+    </ItemGroup>
+
+    <Error Text="System.Private.CoreLib should already be crossgen compiled."
+           Condition=" '%(FilesToCrossGen.Filename)' == 'System.Private.CoreLib' " />
+  </Target>
```

Comment:
Can you change the logic so that System.Private.CoreLib is not required to be already crossgen compiled? We will soon allow linking of System.Private.CoreLib and post-link System.Private.CoreLib will have to be cross-gened.

Reply:
These PRs enable linking of SPC: https://github.com/mono/linker/pull/212, https://github.com/mono/linker/pull/211, https://github.com/dotnet/coreclr/pull/15525.